rust/main/chains/hyperlane-aleo/src/indexer/interchain_gas.rs

```diff
@@ -0,0 +1,141 @@
+    HyperlaneProvider, Indexed, Indexer, InterchainGasPaymaster, InterchainGasPayment, LogMeta,
+    SequenceAwareIndexer, H256, H512,
+};
+
+use crate::{
+    indexer::AleoIndexer, AleoInterchainGasPaymaster, AleoProvider, ConnectionConf, CurrentNetwork,
+    GasPaymentEvent, HookEventIndex, HyperlaneAleoError,
+};
+
+/// Aleo InterchainGas Indexer
+#[derive(Debug, Clone)]
+pub struct AleoInterchainGasIndexer {
+    client: AleoProvider,
+    address: H256,
+    program: String,
+    aleo_address: Address<CurrentNetwork>,
+    domain: HyperlaneDomain,
+}
+
+impl AleoInterchainGasIndexer {
+    /// Creates a new IGP Indexer
+    pub fn new(
+        provider: AleoProvider,
+        locator: &ContractLocator,
+        conf: &ConnectionConf,
+    ) -> ChainResult<Self> {
+        let aleo_address = Address::<CurrentNetwork>::from_bytes_le(locator.address.as_bytes())
+            .map_err(HyperlaneAleoError::from)?;
+        Ok(Self {
+            client: provider,
+            address: locator.address,
+            program: conf.hook_manager_program.clone(),
+            aleo_address,
+            domain: locator.domain.clone(),
+        })
+    }
+}
+
+impl HyperlaneChain for AleoInterchainGasIndexer {
+    /// Return the domain
+    fn domain(&self) -> &HyperlaneDomain {
+        &self.domain
+    }
+
+    /// A provider for the chain
+    fn provider(&self) -> Box<dyn HyperlaneProvider> {
+        Box::new(self.client.clone())
+    }
+}
+
+impl HyperlaneContract for AleoInterchainGasIndexer {
+    /// Address
+    fn address(&self) -> H256 {
+        self.address
+    }
+}
+
+impl InterchainGasPaymaster for AleoInterchainGasIndexer {}
+
+impl AleoIndexer for AleoInterchainGasIndexer {
+    const INDEX_MAPPING: &str = "last_event_index";
+    const VALUE_MAPPING: &str = "gas_payment_events";
+
+    type AleoType = GasPaymentEvent;
+    type Type = InterchainGasPayment;
+
+    fn get_client(&self) -> &AleoProvider {
+        &self.client
+    }
+
+    fn get_program(&self) -> &str {
+        &self.program
+    }
+
+    /// Returns the latest event index of that specific block
+    async fn get_latest_event_index(&self, height: u32) -> ChainResult<u32> {
+        let key = HookEventIndex {
+            hook: self.aleo_address,
+            block_height: height,
+        };
+        // The latest event index for hooks is composition of block_height & hook_address
+        let last_event_index: u32 = self
+            .get_client()
+            .get_mapping_value(self.get_program(), Self::INDEX_MAPPING, &key)
+            .await?;
+        Ok(last_event_index)
+    }
+
+    /// Returns the event value of a mapping
+    fn get_mapping_key<N: Network>(&self, index: u32) -> ChainResult<Plaintext<N>> {
+        let str_value = format!("{{hook: {}, index: {}u32}}", self.aleo_address, index);
+        Ok(Plaintext::from_str(&str_value).map_err(HyperlaneAleoError::from)?)
```

Comment:
This implementation is the same as for Merkle tree hook. I wonder if it is correct? If yes, can we avoid code repetition?